### PR TITLE
Refactor AR distribution

### LINKED ISSUE #5291: Add autoregressive model to V4
created this issue linked to the project for porting time series distributions to v4

### PR DESCRIPTION
**Pull request by ricardoV94**
* Adds random method
* ~~Adds arbitrary noise for the innovations~~ Not yet
* Batched dimensions are now on the left (like other RVs)
* More extensive tests for batching
* Removes AR1 class
  * Could not figure out why `tau_e` is specified as it is, and in general it lacked a good docstring and other kwargs. We can always reintroduce it, but maybe it's useful to first deprecate because of `tau_e`

Closes #5291
Supersedes #5388

Also considered relying on Aeppl, but the returned Scan based logp performs extremely slowly: https://gist.github.com/ricardoV94/2167ab7214affef47a86582141205bf5

### CODE REVIEW

File: pymc/distributions/timeseries.py [hunk 1]
```diff
@@ -40,6 +53,34 @@
 ]
 
 
+def get_steps_from_shape(steps, shape, step_shape_offset=0):
+    """Extract number of steps from shape information
+
+    `step_shape_offset` can be used when the number of steps and shape are not identical
```
Comment (twiecki) [inline pymc/distributions/timeseries.py]: Add argument description.

File: pymc/distributions/timeseries.py [hunk 2]
```diff
@@ -243,142 +266,289 @@ def logp(
...
+            init_dist = kwargs["init"]
+
+        steps = get_steps_from_shape(steps, kwargs.get("shape", None))
+        steps = at.as_tensor_variable(intX(steps), ndim=0)
+
+        if ar_order is None:
+            # TODO: Test case where this fails and passing ar_order directly solves it
+            # We need to know the order of the AR. We perform constant_folding here
+            # to retrieve them
+            shape_fg = FunctionGraph(
```
Comment (twiecki) [inline pymc/distributions/timeseries.py:394]: Maybe add a comment as to what's happening here, hard to see.

File: pymc/distributions/timeseries.py [hunk 3]
```diff
@@ -243,142 +266,289 @@ def logp(
...
-                x = at.add(
-                    *(self.rho[i] * value[self.p - (i + 1) : -(i + 1)] for i in range(self.p))
-                )
-            eps = value[self.p :] - x
+                mu = at.sum(prev_xs * reversed_rhos, axis=0)
+            next_rng, new_x = Normal.dist(mu=mu, sigma=sigma, rng=rng).owner.outputs
+            return new_x, {rng: next_rng}
+
+        # We transpose inputs as scan iterates over first dimension
+        innov_, innov_updates_ = aesara.scan(
```
Comment (twiecki) [inline pymc/distributions/timeseries.py:495]: Great to see this being used, but wasn't this super slow still?

Comment (ricardoV94) [reply pymc/distributions/timeseries.py:495]: This is for generating random draws, which should not be as time-critical. It was the Scan-based logp in NUTS that was painfully slow

**Comment by twiecki** [comment]
This is pretty intense, I could have never written that. I suppose the complexity is somewhat warranted because the distribution has to handle multiple orders and batch dimensions so I'd expect there to be some shape complexity. But I suppose my hope was that with `scan` this could look much simpler.

Anyway, that's more of a general statement as to what it takes to create a new distribution (which I think @canyon289 also expressed), not a comment on this PR which I trust could not be simplified.

File: pymc/distributions/timeseries.py (hunk 3 above)
Comment (twiecki) [reply pymc/distributions/timeseries.py:495]: Oh I see, this is only for sampling?

**Comment by twiecki** [comment]
> Also considered relying on Aeppl, but the returned Scan based logp performs extremely slowly: https://gist.github.com/ricardoV94/2167ab7214affef47a86582141205bf5

Is there an issue about this? Also, curious if this would have made things much simpler?

**Comment by ricardoV94** [comment]
> This is pretty intense, I could have never written that. I suppose the complexity is somewhat warranted because the distribution has to handle multiple orders and batch dimensions so I'd expect there to be some shape complexity. But I suppose my hope was that with `scan` this could look much simpler.
> 
> Anyway, that's more of a general statement as to what it takes to create a new distribution (which I think @canyon289 also expressed), not a comment on this PR which I trust could not be simplified.

The reason this is complicated, is that we are not talking about a single distribution but more like a self-contained model-factory. The extra logic is there to 1) coordinate different distributions and 2) make life easier for users so that they don't need to create in advance the building blocks with the right shapes. 

Once we have it in place it is also easy to extend. For instance, it is now pretty straightforward to allow for arbitrary noise distributions instead of only Normal, by letting users to pass a `noise_dist` like they can do with `init_dist`, and plug it in the appropriate places.

**Comment by ricardoV94** [comment]
> > Also considered relying on Aeppl, but the returned Scan based logp performs extremely slowly: https://gist.github.com/ricardoV94/2167ab7214affef47a86582141205bf5
> Is there an issue about this? Also, curious if this would have made things much simpler?

Not really, it would only save us having to write the logp method ourselves. We still need the logic to generate the random method and do automatic resizing, which is where the new complexity comes from. This was completely missing in V3!

For the logp performance question, there is a somewhat related issue here: https://github.com/aesara-devs/aesara/issues/617

File: pymc/distributions/timeseries.py (hunk 3 above)
Comment (ricardoV94) [reply pymc/distributions/timeseries.py:495]: Yup

File: pymc/distributions/timeseries.py (hunk 2 above)
Comment (ricardoV94) [reply pymc/distributions/timeseries.py:394]: Done

File: pymc/distributions/timeseries.py (hunk 1 above)
Comment (twiecki) [reply pymc/distributions/timeseries.py]: After this LGTM.

Comment (ricardoV94) [reply pymc/distributions/timeseries.py]: Done

File: pymc/distributions/timeseries.py [hunk 4]
```diff
@@ -243,142 +285,294 @@ def logp(
...
+        ar_op = AutoRegressiveRV(
+            inputs=[rhos_, sigma_, init_, steps_],
+            outputs=[innov_next_rng_, ar_],
+            ar_order=ar_order,
+            constant_term=constant_term,
+            inline=True,
+        )
+
+        ar = ar_op(rhos, sigma, init_dist, steps)
+        noise_rng.default_update = ar.owner.outputs[0]
```
Comment (ricardoV94) [inline pymc/distributions/timeseries.py]: This will be changed in #5738, or here, depending on what gets merged first

**Comment by canyon289** [comment]
> Once we have it in place it is also easy to extend. For instance, it is now pretty straightforward to allow for arbitrary noise distributions instead of only Normal, by letting users to pass a noise_dist like they can do with init_dist, and plug it in the appropriate places.

In the following comment I mean this with complete respect, its easy for you to extend Ricardo because you're the man genius that knows whats going on. What would help new contributors a lot are comments, type hints in everything, including tests, and even for lines of code that look trivial. 

I didnt want to bug you too much but if the above sounds good I could indicate everywhere that I personally could use a comment to understand whats going on, would that be helpful?

**Comment by ricardoV94** [comment]
> I didnt want to bug you too much but if the above sounds good I could indicate everywhere that I personally could use a comment to understand whats going on, would that be helpful?

Perhaps take a look first at how Mixtures are implemented, it's basically the same process as in here. Seeing the commonalities might help you understand what is going on: https://github.com/pymc-devs/pymc/blob/main/pymc/distributions/mixture.py

**Comment by canyon289** [comment]
> > I didnt want to bug you too much but if the above sounds good I could indicate everywhere that I personally could use a comment to understand whats going on, would that be helpful?
> 
> Perhaps take a look first at how Mixtures are implemented, it's basically the same process as in here. Seeing the commonalities might help you understand what is going on: https://github.com/pymc-devs/pymc/blob/main/pymc/distributions/mixture.py

I certainly will ! Be prepared that I might start requesting docstrings there as well!

Here's where I'm coming from
Nicoleta will (likely) join us for GSOC soon. When she pulls up this code how will she know to look to mixtures to understand whats going on from the code here? And then there'lll be a dev after that and another dev.

PyMC is an awesome project and its going to be around for years so in addition to building great functionality now, itll be good to ensure

I'm not saying you havent done this btw, I havent done a review yet as Ive been awake for 20 mins, just responding to the comments

**Comment by canyon289** [comment]
@ricardoV94 is it fine if I review on Saturday my time? I really want to do a thorough job, both for my learning and so I can support the Time Series GSOC project

Thats my side of things , let me know what you "i want this merged by X date" timeline is and I'll work around that :)

**Comment by ricardoV94** [comment]
Sure, I am in no rush

**Comment by ricardoV94** [comment]
> Nicoleta will (likely) join us for GSOC soon. When she pulls up this code how will she know to look to mixtures to understand what's going on from the code here? And then there will be a dev after that and another dev.

They will have to read the code and check documentation of the functions used, and if that's not enough set up a debugger and start breaking the code.

Taking to the extreme we can't include a whole Aesara tutorial as comments for future devs. If the same functionality can be achieved by simpler more maintainable means that's a different question.

**Comment by canyon289** [comment]
> Taking to the extreme we can't include a whole Aesara tutorial as comments for future devs. If the same functionality can be achieved by simpler more maintainable means that's a different question.

Agree we cant take this to the extreme, just exploring if there is an optimum between the two

> They will have to read the code and check documentation of the functions used, and if that's not enough set up a debugger and start breaking the code.

Normally id agree but realize this also is more difficult than normal code  due to the way aesara and pymc is integrated. The code isn't "run directly" but sort of sucked into Aesara and recompiled and goes through a number of transformational steps that make it difficult to trace whats exactly going on. 

Anyhow this is stuff thats outside of the scope of this PR, overall really excited to keep building out great functionality, and youre always quite kind and generous in mentoring others ❤️ 

Looking forward to reviewing this and asking a bunch of questions!

**Review by twiecki — APPROVED** [review]

**Comment by ricardoV94** [comment]
~~Will update after #5738~~ Updated

File: pymc/distributions/timeseries.py [hunk 5]
```diff
@@ -11,27 +11,40 @@
...
-from typing import Tuple, Union
+from typing import Optional, Tuple, Union
 
+import aesara
 import aesara.tensor as at
 import numpy as np
 
+from aeppl.abstract import (
+    MeasurableVariable,
+    _get_measurable_outputs,
```
Comment (canyon289) [inline pymc/distributions/timeseries.py]: This is an aeppl comment, if `_get_measurable_outputs` is indeed implicitly private, then we shouldnt import it here. If its meant to be imported we should remove the underscore. 

Its a minor thing but it helps with readability and code understanding. Same with `_logprob` below

Comment (canyon289) [reply pymc/distributions/timeseries.py]: Thats outside the scope of this PR, just noting if we agree issue tickets can be created to track

File: pymc/distributions/timeseries.py [hunk 6]
```diff
@@ -247,142 +289,298 @@ def logp(
...
-       effect of lagged value on current value
-    tau_e: tensor
-       precision for innovations
-    """
+    default_output = 1
+    ar_order: int
+    constant_term: bool
 
-    def __init__(self, k, tau_e, *args, **kwargs):
+    def __init__(self, *args, ar_order, constant_term, **kwargs):
```
Comment (canyon289) [inline pymc/distributions/timeseries.py:294]: Can we add a docstring?

File: pymc/distributions/timeseries.py [hunk 7]
```diff
@@ -247,142 +289,298 @@ def logp(
...
         super().__init__(*args, **kwargs)
-        self.k = k = at.as_tensor_variable(k)
-        self.tau_e = tau_e = at.as_tensor_variable(tau_e)
-        self.tau = tau_e * (1 - k**2)
-        self.mode = at.as_tensor_variable(0.0)
 
-    def logp(self, x):
-        """
-        Calculate log-probability of AR1 distribution at specified value.
+    def update(self, node: Node):
```
Comment (canyon289) [inline pymc/distributions/timeseries.py:299]: Given theres a code comment can this be change to a proper docstring as well

File: pymc/distributions/timeseries.py [hunk 8]
```diff
@@ -247,142 +289,298 @@ def logp(
...
+        the coefficient for the n-th lag.
+    sigma: tensor_like of float, optional
+        Standard deviation of innovation (sigma > 0). Defaults to 1. Only required if
+        tau is not specified.
+    tau: tensor_like of float
+        Precision of innovation (tau > 0).
+    constant: bool, optional
+        Whether the first element of rho should be used as a constant term in the AR
+        process. Defaults to False
+    init_dist: unnamed distribution, optional
```
Comment (canyon289) [inline pymc/distributions/timeseries.py:333]: Does unnamed mean user selected distribution? Oh or one not named in PyMC

File: pymc/distributions/timeseries.py [hunk 9]
```diff
@@ -247,142 +289,298 @@ def logp(
...
+        Scalar or vector distribution for initial values. Defaults to Normal(0, sigma).
+        Distribution should be  created via the `.dist()` API, and have  dimension
+        (*size, ar_order). If not, it will be automatically resized.
+
+        .. warning:: init_dist will be cloned, rendering it independent of the one passed as input.
```
Comment (canyon289) [inline pymc/distributions/timeseries.py:338]: Great warning

File: pymc/distributions/timeseries.py [hunk 10]
```diff
@@ -247,142 +289,298 @@ def logp(
...
+    .. code-block:: python
 
-        if isinstance(rho, list):
-            p = len(rho)
-        else:
-            try:
-                shape_ = rho.shape.tag.test_value
-            except AttributeError:
-                shape_ = rho.shape
+        # Create an AR of order 3, with a constant term
```
Comment (canyon289) [inline pymc/distributions/timeseries.py:353]: Great example

File: pymc/distributions/timeseries.py [hunk 11]
```diff
@@ -247,142 +289,298 @@ def logp(
...
+        with pm.Model() as AR3:
+            coefs = pm.Normal("coefs", 0, size=4)
+            init = pm.Normal.dist(5, size=3)
```
Comment (canyon289) [inline pymc/distributions/timeseries.py:358]: Does size=3  need to match the 3 in AR3? If so add a comment saying they must match so people know

File: pymc/distributions/timeseries.py [hunk 12]
```diff
@@ -247,142 +289,298 @@ def logp(
...
+        if ar_order is None:
+            # If ar_order is not specified we do constant folding on the shape of rhos
+            # to retrieve it. For example, this will detect that
+            # Normal(size=(5, 3)).shape[-1] == 3, which is not known by Aesara before.
+            shape_fg = FunctionGraph(
+                outputs=[rhos.shape[-1]],
+                features=[ShapeFeature()],
+                clone=True,
+            )
+            (folded_shape,) = optimize_graph(shape_fg, custom_opt=topo_constant_folding).outputs
```
Comment (canyon289) [inline pymc/distributions/timeseries.py:399]: Nice optimization

File: pymc/distributions/timeseries.py [hunk 13]
```diff
@@ -247,142 +289,298 @@ def logp(
...
+                    "explictily or make sure rho have a static shape"
+                )
+            ar_order = int(folded_shape) - int(constant)
+            if ar_order < 1:
+                raise ValueError(
+                    "Inferred ar_order is smaller than 1. Increase the last dimension "
+                    "of rho or remove constant_term"
+                )
+
+        if init_dist is not None:
```
Comment (canyon289) [inline pymc/distributions/timeseries.py:413]: Is this something we should make a utility function for all time series eventually? No need on this PR just asking

File: pymc/distributions/timeseries.py [hunk 14]
```diff
@@ -247,142 +289,298 @@ def logp(
...
+            if not isinstance(init_dist, TensorVariable) or not isinstance(
+                init_dist.owner.op, RandomVariable
+            ):
+                raise ValueError(
+                    f"Init dist must be a distribution created via the `.dist()` API, "
+                    f"got {type(init_dist)}"
+                )
+                check_dist_not_registered(init_dist)
+            if init_dist.owner.op.ndim_supp > 1:
```
Comment (canyon289) [inline pymc/distributions/timeseries.py:422]: Thank you for adding all these very helpful exceptions for users

File: pymc/distributions/timeseries.py [hunk 15]
```diff
@@ -247,142 +289,298 @@ def logp(
...
         else:
-            self.p = p
+            # Sigma must broadcast with ar_order
+            init_dist = Normal.dist(sigma=at.shape_padright(sigma), size=(*sigma.shape, ar_order))
 
-        self.constant = constant
-        self.rho = rho = at.as_tensor_variable(rho)
-        self.init = init or Flat.dist()
+        # Tell Aeppl to ignore init_dist, as it will be accounted for in the logp term
+        init_dist = assign_custom_measurable_outputs(init_dist.owner).default_output()
```
Comment (canyon289) [inline pymc/distributions/timeseries.py]: In the GRW pr we did something like `init.tag.ignore_logprob = True`

I see thats now replaced by this function 
https://github.com/pymc-devs/pymc/blob/a3bd08392864adf2ac2c0143367f580b89cf235c/pymc/distributions/logprob.py#L286

Just curious why the change?


image

File: pymc/distributions/timeseries.py [hunk 16]
```diff
@@ -247,142 +289,298 @@ def logp(
...
-        """
-        Calculate log-probability of AR distribution at specified value.
+        return super().dist([rhos, sigma, init_dist, steps, ar_order, constant], **kwargs)
 
-        Parameters
-        ----------
-        value: numeric
-            Value for which log-probability is calculated.
+    @classmethod
+    def num_rngs(cls, *args, **kwargs):
```
Comment (canyon289) [inline pymc/distributions/timeseries.py:437]: Please add docstring for classmethod to explain magic number

File: pymc/distributions/timeseries.py (hunk 5 above)
Comment (ricardoV94) [reply pymc/distributions/timeseries.py]: We are using `_x` and `x` for dispatched functions in Aeppl and PyMC. The underscore one is the basic dispatch function, with which you register on new types, and the one without underscore is the user friendly one that works as a wrapper around the underscored ones. E.g., users call `pm.logcdf(rv, value)` which does a couple of things before calling `_logcdf(rv.owner.op, value, *rv.owner.inputs)`...

File: pymc/distributions/timeseries.py [hunk 17]
```diff
@@ -247,142 +289,298 @@ def logp(
...
-        -------
-        TensorVariable
-        """
-        if self.constant:
-            x = at.add(
-                *(self.rho[i + 1] * value[self.p - (i + 1) : -(i + 1)] for i in range(self.p))
-            )
-            eps = value[self.p :] - self.rho[0] - x
+    @classmethod
+    def ndim_supp(cls, *args):
```
Comment (canyon289) [inline pymc/distributions/timeseries.py:441]: Please add docstring to explain this as well. I think I get it, though a short docstring would reduce friction for those who dont know or in case I forgot :)

File: pymc/distributions/timeseries.py [hunk 18]
```diff
@@ -247,142 +289,298 @@ def logp(
...
+            inline=True,
+        )
+
+        ar = ar_op(rhos, sigma, init_dist, steps)
+        return ar
 
-        innov_like = Normal.dist(mu=0.0, tau=self.tau).logp(eps)
-        init_like = self.init.logp(value[: self.p])
+    @classmethod
+    def change_size(cls, rv, new_size, expand=False):
```
Comment (canyon289) [inline pymc/distributions/timeseries.py:517]: Add docstring, especially for expand

File: pymc/distributions/timeseries.py [hunk 19]
```diff
@@ -247,142 +289,298 @@ def logp(
...
+        )
 
-        return at.sum(innov_like) + at.sum(init_like)
+
+MeasurableVariable.register(AutoRegressiveRV)
+
+
+@_get_measurable_outputs.register(AutoRegressiveRV)
+def _get_measurable_outputs_ar(op, node):
+    # This tells Aeppl that the second output is the measurable one
```
Comment (canyon289) [inline pymc/distributions/timeseries.py:541]:
```suggestion
    """This tells Aeppl that the second output is the measurable one."""
```

File: pymc/distributions/timeseries.py [hunk 20]
```diff
@@ -247,142 +289,298 @@ def logp(
...
+
+    ar_order = op.ar_order
+    constant_term = op.constant_term
+
+    # Convolve rhos with values
+    if constant_term:
+        expectation = at.add(
+            rhos[..., 0, None],
+            *(
+                rhos[..., i + 1, None] * value[..., ar_order - (i + 1) : -(i + 1)]
```
Comment (canyon289) [inline pymc/distributions/timeseries.py:557]: Fancy

File: pymc/distributions/timeseries.py (hunk 15 above)
Comment (ricardoV94) [reply pymc/distributions/timeseries.py]: Aeppl changed and no longer cares about that tag. The GRW was updated in this recent PR: https://github.com/pymc-devs/pymc/pull/5738

File: pymc/distributions/timeseries.py [hunk 21]
```diff
@@ -247,142 +289,298 @@ def logp(
...
+    )
+    init_logp = logp(init_dist, value[..., :ar_order])
+    if init_dist.owner.op.ndim_supp == 0:
+        init_logp = at.sum(init_logp, axis=-1)
+    return init_logp + innov_logp
+
+
+@_moment.register(AutoRegressiveRV)
+def ar_moment(op, rv, rhos, sigma, init_dist, steps, noise_rng):
+    # Use last entry of init_dist moment as the moment for the whole AR
```
Comment (canyon289) [inline pymc/distributions/timeseries.py:580]:
```suggestion
    """Use last entry of init_dist moment as the moment for the whole AR."""
```

File: pymc/distributions/timeseries.py (hunk 16 above)
Comment (ricardoV94) [reply pymc/distributions/timeseries.py:437]: It's clearly explained in the base SymbolicDistribution

File: pymc/distributions/timeseries.py (hunk 17 above)
Comment (ricardoV94) [reply pymc/distributions/timeseries.py:441]: Same

File: pymc/tests/test_distributions_timeseries.py [hunk 22]
```diff
@@ -143,60 +142,198 @@ def test_gaussian_random_walk_init_dist_logp(self, init):
...
+            y = AR("y", phi, sigma=1, init_dist=Flat.dist(), shape=len(data))
+            z = Normal(
+                "z", mu=phi[0] * data[1:-1] + phi[1] * data[:-2], sigma=1, shape=len(data) - 2
+            )
+        ar_like = t.compile_logp(y)({"y": data})
+        reg_like = t.compile_logp(z)({"z": data[2:]})
+        np.testing.assert_allclose(ar_like, reg_like)
+
+    @pytest.mark.parametrize("constant", (False, True))
+    def test_batched_size(self, constant):
```
Comment (canyon289) [inline pymc/tests/test_distributions_timeseries.py:205]: Add docstring, its bordering obvious but its not obvious enough that I think a newcomer would get it

File: pymc/tests/test_distributions_timeseries.py [hunk 23]
```diff
@@ -143,60 +142,198 @@ def test_gaussian_random_walk_init_dist_logp(self, init):
...
-        rho = Normal("rho", 0.0, 1.0)
-        y1 = AR1("y1", rho, 1.0, observed=data)
-        y2 = AR("y2", rho, 1.0, init=Normal.dist(0, 1), observed=data)
-    initial_point = t.initial_point()
-    np.testing.assert_allclose(y1.logp(initial_point), y2.logp(initial_point))
+        y_eval = draw(y, draws=2)
+        assert y_eval[0].shape == (batch_size, steps)
+        assert not np.any(np.isclose(y_eval[0], y_eval[1]))
+
+    def test_batched_rhos(self):
```
Comment (canyon289) [inline pymc/tests/test_distributions_timeseries.py:226]: Same comment as above

File: pymc/tests/test_distributions_timeseries.py [hunk 24]
```diff
@@ -143,60 +142,198 @@ def test_gaussian_random_walk_init_dist_logp(self, init):
...
-        z = Normal("z", mu=phi[0] * data[1:-1] + phi[1] * data[:-2], sigma=1, shape=len(data) - 2)
-    ar_like = t["y"].logp({"z": data[2:], "y": data})
-    reg_like = t["z"].logp({"z": data[2:], "y": data})
-    np.testing.assert_allclose(ar_like, reg_like)
+        beta_tp[1] = 0  # Should always be close to zero
+        y_eval = t0["y"].eval({t0["beta"]: beta_tp})
+        assert y_eval.shape == (batch_size, steps)
+        assert np.all(abs(y_eval[1]) < 5)
+
+    def test_batched_sigma(self):
```
Comment (canyon289) [inline pymc/tests/test_distributions_timeseries.py:248]: Same for all batched tests

**Review by canyon289 — APPROVED** [review]
This is awesome. Most comments are essentially add more docstrings on internal functions to aid future developers, or forgetful developers.

File: pymc/distributions/timeseries.py (hunk 5 above)
Comment (canyon289) [reply pymc/distributions/timeseries.py]: Oh I see, makes sense then, thanks for explanation!

File: pymc/distributions/timeseries.py (hunk 15 above)
Comment (canyon289) [reply pymc/distributions/timeseries.py]: Cool!

File: pymc/distributions/timeseries.py (hunk 6 above)
Comment (ricardoV94) [reply pymc/distributions/timeseries.py:294]: What would it say? This init method clearly saves two variables as class instance properties and sends the rest to the super class.

File: pymc/distributions/timeseries.py (hunk 16 above)
Comment (canyon289) [reply pymc/distributions/timeseries.py:437]: I cant find the string `num_rngs` anywhere in SymbolicDistribution or `distribution,py`

https://github.com/pymc-devs/pymc/blob/main/pymc/distributions/distribution.py

image

File: pymc/distributions/censored.py [hunk 25]
```diff
@@ -76,6 +76,14 @@ def dist(cls, dist, lower, upper, **kwargs):
         check_dist_not_registered(dist)
         return super().dist([dist, lower, upper], **kwargs)
 
+    @classmethod
+    def num_rngs(cls, *args, **kwargs):
+        return 1
+
+    @classmethod
+    def ndim_supp(cls, *dist_params):
```
Comment (canyon289) [inline pymc/distributions/censored.py:84]:
```suggestion
    def ndim_supp(cls, *dist_params):
        """Refer to base class SymbolicDistribution for documentation."""
```

File: pymc/distributions/censored.py [hunk 26]
```diff
@@ -76,6 +76,14 @@ def dist(cls, dist, lower, upper, **kwargs):
         check_dist_not_registered(dist)
         return super().dist([dist, lower, upper], **kwargs)
 
+    @classmethod
+    def num_rngs(cls, *args, **kwargs):
```
Comment (canyon289) [inline pymc/distributions/censored.py:80]:
```suggestion
    def num_rngs(cls, *args, **kwargs):
        """Refer to base class SymbolicDistribution for documentation."""
```

File: pymc/distributions/timeseries.py (hunk 17 above)
Comment (canyon289) [reply pymc/distributions/timeseries.py:441]: Added a docstring suggestion

File: pymc/distributions/timeseries.py (hunk 16 above)
Comment (canyon289) [reply pymc/distributions/timeseries.py:437]: Oh nevermind its in this PR, sorry, reviewing on a single screen laptop which I typically dont do. My apologies. See commit suggestion below which lets folks know to refer to symbolic dist

File: pymc/distributions/distribution.py [hunk 27]
```diff
@@ -396,18 +396,19 @@ def __new__(
             to a canonical parametrization. It should call `super().dist()`, passing a
             list with the default parameters as the first and only non keyword argument,
             followed by other keyword arguments like size and rngs, and return the result
+        cls.num_rngs
+            Returns the number of rngs given the same arguments passed by the user when
+            calling the distribution
+        cls.ndim_supp
+            Returns the support of the symbolic distribution, given the default set of
+            parameters. This may not always be constant, for  instance if the symbolic
```
Comment (canyon289) [inline pymc/distributions/distribution.py]:
```suggestion
            parameters. This may not always be constant, for instance if the symbolic
```

File: pymc/distributions/distribution.py [hunk 28]
```diff
@@ -396,18 +396,19 @@ def __new__(
             to a canonical parametrization. It should call `super().dist()`, passing a
             list with the default parameters as the first and only non keyword argument,
             followed by other keyword arguments like size and rngs, and return the result
+        cls.num_rngs
```
Comment (canyon289) [inline pymc/distributions/distribution.py]: Really nice docstrings!

File: pymc/distributions/timeseries.py (hunk 6 above)
Comment (canyon289) [reply pymc/distributions/timeseries.py:294]: Ah nothing , I see I just missed it above. Thanks for pushing back

**Comment by canyon289** [comment]
Thanks Ricardo and nicely done. Definitely ready for a merge and if theres anything I still have questions, or think we need docstrings which Im clearly a fan of, I can do it outside of this PR

File: pymc/distributions/timeseries.py (hunk 15 above)
Comment (ricardoV94) [reply pymc/distributions/timeseries.py]: Actually this is outdated already after #5738

File: pymc/distributions/timeseries.py (hunk 13 above)
Comment (ricardoV94) [reply pymc/distributions/timeseries.py:413]: Yeah something like `check_valid_dist(dist, allowed_ndim_supp=(x, y))`, as this is used in all the distributions that take other distributions as inputs. Want to open an issue/PR for that?

File: pymc/distributions/timeseries.py (hunk 8 above)
Comment (ricardoV94) [reply pymc/distributions/timeseries.py:333]: Hmm, I don't like this naming very much. It's supposed to be something like `Normal.dist(mu, sigma)`. I went with unnamed because that's a salient difference from normal calls (like `pm.Normal("name", mu, sigma)`, other than the fact that it uses the `.dist` classmethod of course. The difference for us, is that it is not tracked at the model level/ sampled.

File: pymc/distributions/timeseries.py (hunk 11 above)
Comment (ricardoV94) [reply pymc/distributions/timeseries.py:358]: Done

File: pymc/distributions/timeseries.py (hunk 19 above)
Comment (ricardoV94) [reply pymc/distributions/timeseries.py:541]: I don't think this should be a docstring... but I can't quite put it to words why

File: pymc/distributions/timeseries.py (hunk 21 above)
Comment (ricardoV94) [reply pymc/distributions/timeseries.py:580]: Same here

File: pymc/tests/test_distributions_timeseries.py (hunk 22 above)
Comment (ricardoV94) [reply pymc/tests/test_distributions_timeseries.py:205]: What do you think a useful docstring would be?

File: pymc/distributions/timeseries.py (hunk 7 above)
Comment (ricardoV94) [reply pymc/distributions/timeseries.py:299]: Done

**Comment by canyon289** [comment]
@ricardoV94 I saw your comments but wont be able to give you proper replies for a couple of days. There's nothing that should hold up a merge. Please feel free to do so, but keep your branch around so I can come back to this PR later and refer to the comments to make proposals in separate issues or PRs 🎉

**Comment by ricardoV94** [comment]
Thanks for reviewing @twiecki and @canyon289
